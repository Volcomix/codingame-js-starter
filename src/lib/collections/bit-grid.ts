export class BitGrid {
  private buffer: Buffer
  private offset: number
  private lastBitPosition: BitPosition = { byteIndex: 0, bitIndexInByte: 0 }

  constructor(readonly width: number, readonly height: number) {
    this.offset = Math.ceil(Math.log2(width))
    this.buffer = Buffer.alloc(((height << this.offset) >> 3) + 1)
  }

  reset() {
    this.buffer.fill(0)
  }

  get(x: number, y: number): boolean {
    const { byteIndex, bitIndexInByte } = this.getBitPosition(x, y)
    return ((this.buffer[byteIndex] >> bitIndexInByte) & 1) === 1
  }

  set(x: number, y: number, value: boolean) {
    const { byteIndex, bitIndexInByte } = this.getBitPosition(x, y)
    this.buffer[byteIndex] = value
      ? this.buffer[byteIndex] | (1 << bitIndexInByte)
      : this.buffer[byteIndex] & ~(1 << bitIndexInByte)
  }

  private getBitPosition(x: number, y: number): BitPosition {
    const bitIndex = (y << this.offset) + x
    const byteIndex = bitIndex >> 3
    this.lastBitPosition.byteIndex = byteIndex
    this.lastBitPosition.bitIndexInByte = bitIndex - (byteIndex << 3)
    return this.lastBitPosition
  }
}

interface BitPosition {
  byteIndex: number
  bitIndexInByte: number
}
