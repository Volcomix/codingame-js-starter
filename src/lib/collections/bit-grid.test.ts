import { BitGrid } from './bit-grid'

describe('BitGrid', () => {
  let bitGrid: BitGrid

  beforeEach(() => {
    bitGrid = new BitGrid(15, 15)
  })

  describe('constructor', () => {
    it('set all cells to false', () => {
      expect(bitGrid.get(1, 1)).toBe(false)
    })
  })

  describe('reset', () => {
    it('set all cells to false', () => {
      bitGrid.set(1, 1, true)
      bitGrid.reset()
      expect(bitGrid.get(1, 1)).toBe(false)
    })
  })

  describe('get', () => {
    it('returns true', () => {
      bitGrid.set(11, 12, true)
      expect(bitGrid.get(11, 12)).toBe(true)
    })

    it('returns false', () => {
      bitGrid.set(1, 1, false)
      expect(bitGrid.get(1, 1)).toBe(false)
    })
  })

  it('matches snapshot', () => {
    bitGrid.set(0, 0, true)
    bitGrid.set(1, 1, true)
    bitGrid.set(14, 14, true)
    expect(bitGrid).toMatchSnapshot()
  })
})
