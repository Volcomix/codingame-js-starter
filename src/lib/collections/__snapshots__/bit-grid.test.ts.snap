// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BitGrid matches snapshot 1`] = `
BitGrid {
  "buffer": Object {
    "data": Array [
      1,
      0,
      2,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      64,
      0,
    ],
    "type": "Buffer",
  },
  "height": 15,
  "lastBitPosition": Object {
    "bitIndexInByte": 6,
    "byteIndex": 29,
  },
  "offset": 4,
  "width": 15,
}
`;
